### DISCUSSION TITLE
502 Bad Gateway Error When Accessing  Remix App version 1.16.0-pre.6 with Nginx

### DISCUSSION THREAD

**Discussion by hakimLyon** [remix-run/remix#6258] [Q&A] [answered]
Description: 
I have configured Nginx to use a reverse proxy to Remix app version  1.16.0-pre.6 that uses an Express server listening on port 3201. I have also configured SSL for my domain ```example.com``` for test. However, when I try to access via ```https://example.com```, I get a ```502 Bad Gateway error```. Accessing the application via ```http://localhost:3201``` or ```http://example:3201``` works correctly.

Here's my Nginx configuration:
```nginx
server {
    listen 80;
    listen [::]:80;

    server_name example.com www.example.com;

    return 301 https://example.com$request_uri;
}

server {
    listen 443 ssl;
    listen [::]:443 ssl;

    server_name example.com;

    ssl_certificate /path/to/ssl/certificate;
    ssl_certificate_key /path/to/ssl/key;

    location / {
        proxy_pass http://localhost:3201;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    }
}
```
Server 
```js
const path = require("path");
const express = require("express");
const compression = require("compression");
const morgan = require("morgan");
const { createRequestHandler } = require("@remix-run/express");
const { installGlobals } = require("@remix-run/node");

installGlobals();

const BUILD_DIR = path.join(process.cwd(), "build");

const app = express();

app.use(compression());

app.disable("x-powered-by");

app.use(
  "/build",
  express.static("public/build", { immutable: true, maxAge: "1y" })
);


app.use(express.static("public", { maxAge: "1h" }));

app.use(morgan("tiny"));

app.all(
  "*",
  process.env.NODE_ENV === "development"
    ? (req, res, next) => {
        purgeRequireCache();

        return createRequestHandler({
          build: require(BUILD_DIR),
          mode: process.env.NODE_ENV,
        })(req, res, next);
      }
    : createRequestHandler({
        build: require(BUILD_DIR),
        mode: process.env.NODE_ENV,
      })
);
const port = process.env.PORT || 3201;

app.listen(port, () => {
  console.log(`Express server listening on port ${port}`);
// in development, call `broadcastDevReady` _after_ your server is up and running
			if (process.env.NODE_ENV === 'development') {
				const { broadcastDevReady } = await import('@remix-run/server-runtime')
				broadcastDevReady(build)
			}
});

function purgeRequireCache() {
  // purge require cache on requests for "server side HMR" this won't let
  // you have in-memory objects between requests in development,
  // alternatively you can set up nodemon/pm2-dev to restart the server on
  // file changes, but then you'll have to reconnect to databases/etc on each
  // change. We prefer the DX of this, so we've included it for you by default
  for (const key in require.cache) {
    if (key.startsWith(BUILD_DIR)) {
      delete require.cache[key];
    }
  }
}
```

I have checked the following:

- The Remix application is accessible from the Nginx server using the `curl` command.
- The domain name example.com points to the IP address of my Nginx server.
- Port 3201 is open on my server.
- The SSL certificate I'm using is correctly configured in the Nginx configuration.

I don't know what else to do to resolve the issue 🤷🏿‍♂️ . Can anyone help?

**Comment by kwiat1990** [comment]
I have the very same code as of today and in yours the `build` variable passed in into `broadcastDevReady` doesn’t exist. I’m also not sure regarding placing import not at the top of the file.

**Reply by hakimLyon** [reply]
Thank you for bringing that to my attention. My apologies for any confusion caused. Regarding the 'build' variable,

 ```diff
 + const { broadcastDevReady } = require('@remix-run/server-runtime');
......
   if (process.env.NODE_ENV === 'development') {
 +   let build = require(BUILD_DIR);
    broadcastDevReady(build);
  }`

```


 I inadvertently forgot to include it when creating the discussion. As for the dynamic import, while it's not the standard practice to place it within a function, it can be a useful approach for certain use cases where the import needs to be loaded conditionally. However, I'll keep in mind that placing imports at the top of the file is generally recommended for clarity and consistency.

**Comment by machour** [comment]
First thing first, is http://example.com:3201 showing your running remix app?

If so, this is purely an nginx problem, if not, it's first a remix problem.

For nginx, I've been using this on all my apps:

```
        proxy_pass http://127.0.0.1:3201;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
```

**Reply by hakimLyon** [reply]
Hey @machour  👋

I tried your Nginx configuration , but unfortunately, I still got a ```502 Bad Gateway``` 😩.

I checked if my ```NestJs``` app was working, which is on port 3101 , and whose nginx configuration has a server name of ```https://sub.example.com``` . And it was working perfectly 👍.

So, on my AWS server, I initialized another Remix project 🚀 by selecting [Express server](https://github.com/remix-run/remix/tree/main/templates/express) during creation for testing . And it works, wow 🤩. The domain ```https://example.com``` works without the 502 error 😁 .

Then I tried [Idie-stack](https://github.com/remix-run/indie-stack) and booom 💥, it works too 🎉.

I also tested  [Blues-stack 🎵](https://github.com/remix-run/blues-stack). But it showed the same errors as with my initial Remix app 🤯. When navigating to ```https://example.com```, the ```502 Bad Gateway``` error reappeared.

I'm trying to understand why the one initiated with an  [Express server](https://github.com/remix-run/remix/tree/main/templates/express) was working while  [Blues-stack 🎵](https://github.com/remix-run/blues-stack). was showing ```502 Bad Gateway``` 😕.

**Comment by hakimLyon** [comment, marked as answer]
Great news! I found a solution 🎉. I needed to add the following lines to my nginx configuration:

```nginx
proxy_buffer_size   128k;
proxy_buffers   4 256k;
proxy_busy_buffers_size   256k;
```
With these changes, the Remix app is now working without the ```502 Bad Gateway``` error 😁.

So, the problem was related to the proxy buffering settings in nginx. By default, nginx buffers responses from the proxied server in memory, and the default buffer size is relatively small. For larger responses, the buffer can overflow, causing the "502 Bad Gateway" error.

By increasing the buffer sizes, we are allowing nginx to handle larger responses without overflowing the buffer, and that's why the Remix app is now working correctly 🚀.
